### PR TITLE
Implement a delocalizing contents manager for GCS

### LINKED ISSUE #242: Support configurable delocalization to GCS
Design under discussion @ https://docs.google.com/document/d/1GbaL1otgswBEOqfte8JO8uQDG_vSEG4a-qcKMEdufLI/edit?ts=5a9f0755#heading=h.vrxfsh4mwyvy

From talking with @rtitle yesterday, we think it will make sense to aim to build this as something deployed by default with Leonardo (though there would be no requirement to utilize it). I can prepare an initial PR for this once there is agreement on the doc.

### PR DESCRIPTION
**Pull request by calbach**
Depends on https://github.com/DataBiosphere/leonardo/pull/258
Resolves #242 

See class documentation on `DelocalizingContentsManager` for an overview. I'd like to add some tests once I've got feedback on the overall approach, but I'm not sure there's an existing hook/harness to run them. Thoughts @rtitle / @helgridly ?

---

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by rtitle** [comment]
We have an automated test suite where we should be able to add a test for this -- I can help with this. Will also review in more detail later (but at a glance it looks great). Thanks for doing this.

**Comment by calbach** [comment]
PTAL

Added tests as well with manual instructions for now.

File: jupyter-docker/delocalizing_contents_manager.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+  # Cache delocalization metadata for 5 minutes. The cache points to a tuple of
+  # (metadata: dict, read_at: datetime), with a key of JSON file path. Negative
+  # lookups are stored here as well with None as metadata (this will be the common
+  # case for any clusters which don't use delocalization).
+  def __init__(self, *args, **kwargs):
+    self.log.info('initializing DelocalizingContentsManager')
+    self.delocalize_metadata = {}
+    # Allows for stubbing in tests.
+    self.file_cmd = ['gsutil', '-q']
```
Comment (rtitle) [inline jupyter-docker/delocalizing_contents_manager.py]: Should we think about specifying `-m` as well?
```
$ gsutil help options

  -m          Causes supported operations (acl ch, acl set, cp, mv, rm, rsync,
              and setmeta) to run in parallel. This can significantly improve
              performance if you are performing operations on a large number of
              files over a reasonably fast network connection.
```

**Comment by calbach** [comment]
Yes, I think it's ready. I just made a few more changes, e.g. file renames.

The one outstanding issue here is that I haven't setup the tests to run automatically. I can advise on this, but since I don't have access to your Jenkins I would probably need to someone else to handle the actual code change.

File: jupyter-docker/jupyter_delocalize.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+METADATA_TTL = timedelta(minutes=5)
+
+class DelocalizingContentsManager(FileContentsManager):
+  """
+  A contents manager allowing for configurable automated delocalization.
+
+  Files will always be persisted to the local Jupyter file system as usual. The
+  decision of whether to try delocalization is driven by the presence of a file
+  named .delocalize.json on the Jupyter file system. The format of this
```
Comment (dmohs) [inline jupyter-docker/jupyter_delocalize.py]: This comment gives me almost all of the details I need, but I had to go back to the doc for one piece—the where. I might reword this to: "... presence of a file named .delocalize.json in the same directory as the notebook file that is the target of the save operation."

Comment (calbach) [reply jupyter-docker/jupyter_delocalize.py]: Good point, thanks. Sent https://github.com/DataBiosphere/leonardo/pull/301

Reworded slightly to capture that it is not only applicable to `save()` operations (also moves, deletes), and also is not restricted to notebooks - could be data files if someone uploaded something and the filename matches the given pattern.
